File: sdks/js/src/data/coerce/toObject.test.ts
jest.mock('../../api/client/data/get', () => jest.fn())
import dataGet from '../../api/client/data/get'

import objectUnderTest from './toObject'
import Value from '../../model/value'

describe('toObject', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })
  describe('value is array', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { array: [] }

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce array to object; incompatible types`)
    })
  })
  describe('value is boolean', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { boolean: true }

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce boolean '${providedValue.boolean}' to object; incompatible types`)
    })
  })
  describe('value is dir', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { dir: 'dummyDir' }

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce dir '${providedValue.dir}' to object; incompatible types`)
    })
  })
  describe('value is file', () => {
    it('should call dataGet w/ expected args', async () => {
      /* arrange */
      const providedNetRef = 'providedNetRef'
      const providedValue = { file: 'dummyFile' }

      const dataGetResult = { text: jest.fn().mockResolvedValue(JSON.stringify({})) }
        ; (dataGet as jest.Mock).mockResolvedValue(dataGetResult)

      /* act */
      await objectUnderTest(
        providedNetRef,
        providedValue
      )

      /* assert */
      expect(dataGet)
        .toBeCalledWith(
          providedNetRef,
          providedValue.file
        )
    })
    describe('dataGet errs', () => {
      it('returns expected result', async () => {
        /* arrange */
        const providedValue = { file: 'dummyFile' }

        const dataGetErr = new Error('dummyErr')
          ; (dataGet as jest.Mock).mockRejectedValue(dataGetErr)

        /* act/assert */
        await expect(objectUnderTest(
          'apiBaseUrl',
          providedValue
        ))
          .rejects
          .toThrow(`unable to coerce file to array; error was ${dataGetErr.message}`)
      })
    })
    describe('dataGet doesn\'t err', () => {
      it('should return expected result', async () => {
        /* arrange */
        const providedValue = { file: 'dummyFile' }

        const objectValue = { prop1: 'prop1Value' }
        const dataGetResult = { text: jest.fn().mockResolvedValue(JSON.stringify(objectValue)) }
          ; (dataGet as jest.Mock).mockResolvedValue(dataGetResult)

        /* act */
        const actualResult = await objectUnderTest(
          'apiBaseUrl',
          providedValue
        )

        /* assert */
        expect(actualResult).toEqual({ object: objectValue })
      })
    })
  })
  describe('value is number', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { number: 2.2 }

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce number to object; incompatible types`)
    })
  })
  describe('value is object', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { object: { prop1: 'prop1Value' } }

      /* act */
      const actualResult = await objectUnderTest(
        'apiBaseUrl',
        providedValue
      )

      /* assert */
      expect(actualResult).toEqual(providedValue)
    })
  })
  describe('value is socket', () => {
    it('should return expected result', async () => {
      /* arrange */
      const providedValue = { socket: 'dummySocket' }

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce socket '${providedValue.socket}' to object; incompatible types`)
    })
  })
  describe('value is string', () => {
    describe('string is JSON', () => {
      describe('JSON is object', () => {
        it('should return expected result', async () => {
          /* arrange */
          const objectValue = { prop1: 'prop1Value' }
          const providedValue = { string: JSON.stringify(objectValue) }

          /* act */
          const actualResult = await objectUnderTest(
            'apiBaseUrl',
            providedValue
          )

          /* assert */
          expect(actualResult).toEqual({ object: JSON.parse(providedValue.string) })
        })
      })
      describe('JSON is array', () => {
        it('returns expected result', async () => {
          /* arrange */
          const providedValue = { string: JSON.stringify([]) }

          /* act/assert */
          await expect(objectUnderTest(
            'apiBaseUrl',
            providedValue
          ))
            .rejects
            .toThrow(`unable to coerce string to object; error was parsed array but expected object`)
        })
      })
      describe('JSON is null', () => {
        it('returns expected result', async () => {
          /* arrange */
          const providedValue = { string: JSON.stringify(null) }

          /* act/assert */
          await expect(objectUnderTest(
            'apiBaseUrl',
            providedValue
          ))
            .rejects
            .toThrow(`unable to coerce string to object; error was parsed null but expected object`)
        })
      })
      describe('JSON is string', () => {
        it('returns expected result', async () => {
          /* arrange */
          const providedValue = { string: JSON.stringify('dummyString') }

          /* act/assert */
          await expect(objectUnderTest(
            'apiBaseUrl',
            providedValue
          ))
            .rejects
            .toThrow(`unable to coerce string to object; error was parsed string but expected object`)
        })
      })
    })
    describe('string isn\'t JSON', () => {
      it('returns expected result', async () => {
        /* arrange */
        const providedValue = { string: 'notValidJsonObject' }

        /* act/assert */
        await expect(objectUnderTest(
          'apiBaseUrl',
          providedValue
        ))
          .rejects
          .toThrow(`unable to coerce string to object; error was Unexpected token o in JSON at position 1`)
      })
    })
  })
  describe('value isnt any of the above', () => {
    it('returns expected result', async () => {
      /* arrange */
      const providedValue = { unknown: 'providedValue' } as Value

      /* act/assert */
      await expect(objectUnderTest(
        'apiBaseUrl',
        providedValue
      ))
        .rejects
        .toThrow(`unable to coerce ${JSON.stringify(providedValue)} to object`)
    })
  })
})

File: sdks/go/data/coerce/toObject_test.go
package coerce

import (
	"fmt"
	"os"

	"github.com/ipld/go-ipld-prime"
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var _ = Context("ToObject", func() {
	Context("Value is nil", func() {
		It("should return expected result", func() {
			/* arrange */
			/* act */
			actualValue, actualErr := ToObject(nil)

			/* assert */
			Expect(actualValue).To(BeNil())
			Expect(actualErr).To(BeNil())
		})
	})
	Context("Value.Array isn't nil", func() {
		It("should return expected result", func() {
			/* arrange */
			providedValue := ipld.Node{
				Array: new([]interface{}),
			}

			/* act */
			actualValue, actualErr := ToObject(providedValue)

			/* assert */
			Expect(actualValue).To(BeNil())
			Expect(actualErr).To(MatchError("unable to coerce array to object: incompatible types"))
		})
	})
	Context("Value.Dir isn't nil", func() {
		It("should return expected result", func() {
			/* arrange */
			providedDir := "dummyValue"
			providedValue := ipld.Node{
				Dir: &providedDir,
			}

			/* act */
			actualValue, actualErr := ToObject(providedValue)

			/* assert */
			Expect(actualValue).To(BeNil())
			Expect(actualErr).To(MatchError("unable to coerce dir 'dummyValue' to object: incompatible types"))
		})
	})
	Context("Value.File isn't nil", func() {
		Context("os.ReadFile errs", func() {
			It("should return expected result", func() {
				/* arrange */
				nonExistentPath := "nonExistent"

				/* act */
				actualValue, actualErr := ToObject(
					ipld.Node{File: &nonExistentPath},
				)

				/* assert */
				Expect(actualValue).To(BeNil())
				Expect(actualErr).To(MatchError("unable to coerce file to object: open nonExistent: no such file or directory"))
			})
		})
		Context("os.ReadFile doesn't err", func() {
			Context("json.Unmarshal errs", func() {
				It("should return expected result", func() {
					/* arrange */

					tmpFile, err := os.CreateTemp("", "")
					if err != nil {
						panic(err)
					}

					filePath := tmpFile.Name()

					/* act */
					actualValue, actualErr := ToObject(
						ipld.Node{File: &filePath},
					)

					/* assert */
					Expect(actualValue).To(BeNil())
					Expect(actualErr).To(MatchError("unable to coerce file to object: unexpected end of JSON input"))
				})
			})
			Context("json.Unmarshal doesn't err", func() {
				It("should return expected result", func() {
					/* arrange */
					mapEntryValue := "mapEntryValue"
					mapEntryKey := "mapKey"
					mapValueJSON := fmt.Sprintf(`{"%s": "%s"}`, mapEntryKey, mapEntryValue)

					tmpFile, err := os.CreateTemp("", "")
					if err != nil {
						panic(err)
					}
					filePath := tmpFile.Name()

					err = os.WriteFile(filePath, []byte(mapValueJSON), 0777)
					if err != nil {
						panic(err)
					}

					mapValue := map[string]interface{}{
						mapEntryKey: mapEntryValue,
					}
					expectedValue := ipld.Node{Object: &mapValue}

					/* act */
					actualValue, actualErr := ToObject(
						ipld.Node{File: &filePath},
					)

					/* assert */
					Expect(actualErr).To(BeNil())
					Expect(*actualValue).To(Equal(expectedValue))
				})
			})
		})
	})
	Context("Value.Number isn't nil", func() {
		It("should return expected result", func() {
			/* arrange */
			providedNumber := 2.2
			providedValue := ipld.Node{
				Number: &providedNumber,
			}

			/* act */
			actualValue, actualErr := ToObject(providedValue)

			/* assert */
			Expect(actualValue).To(BeNil())
			Expect(actualErr).To(MatchError("unable to coerce number '2.2' to object: incompatible types"))
		})
	})
	Context("Value.Object isn't nil", func() {
		It("should return expected result", func() {
			/* arrange */
			providedValue := ipld.Node{
				Object: new(map[string]interface{}),
			}

			/* act */
			actualValue, actualErr := ToObject(providedValue)

			/* assert */
			Expect(actualValue).To(Equal(providedValue))
			Expect(actualErr).To(BeNil())
		})
	})
	Context("Value.String isn't nil", func() {
		Context("json.Unmarshal errs", func() {
			It("should return expected result", func() {
				/* arrange */

				/* act */
				actualValue, actualErr := ToObject(
					ipld.Node{String: new(string)},
				)

				/* assert */
				Expect(actualValue).To(BeNil())
				Expect(actualErr).To(MatchError("unable to coerce string to object: unexpected end of JSON input"))
			})
		})
		Context("json.Unmarshal doesn't err", func() {
			It("should return expected result", func() {
				/* arrange */
				mapEntryValue := "mapEntryValue"
				mapEntryKey := "mapKey"
				mapValueJSON := fmt.Sprintf(`{"%s": "%s"}`, mapEntryKey, mapEntryValue)

				mapValue := map[string]interface{}{
					mapEntryKey: mapEntryValue,
				}
				expectedValue := ipld.Node{Object: &mapValue}

				/* act */
				actualValue, actualErr := ToObject(
					ipld.Node{String: &mapValueJSON},
				)

				/* assert */
				Expect(*actualValue).To(Equal(expectedValue))
				Expect(actualErr).To(BeNil())
			})
		})
	})
	Context("Value.Array,Value.Dir,File,Number,Object,String nil", func() {
		It("should return expected result", func() {
			/* arrange */
			providedValue := ipld.Node{}

			/* act */
			actualValue, actualErr := ToObject(providedValue)

			/* assert */
			Expect(actualValue).To(BeNil())
			Expect(actualErr).To(MatchError("unable to coerce '&{Array:<nil> Boolean:<nil> Dir:<nil> File:<nil> Number:<nil> Object:<nil> Socket:<nil> String:<nil>}' to object"))
		})
	})
})
